### PR TITLE
Gt plt summary customization

### PR DESCRIPTION
**Pull request by ScottFB101**
# Summary

Thank you for contributing to **gt-extras**! To make this process easier for everyone, please explain the context and purpose of your contribution. Also, list the changes made to the existing code or documentation.

This PR addresses the 4 out of the 5 enhancements proposed in Issue #146. The one enhancement not addressed was: "Display times instead of just dates". I felt that was deserving of another separate PR. 

The changes made to the existing code are rather straightforward, and don't need much background besides knowing they address the 4 enhancements previously referenced. This is my opinion of course, and I'm happy to elaborate more on the code here if there's a want for it.

As for the context and purpose of my contribution, I use so many posit-backed packages and tools, that it only feels right to get involved in contributing to the ecosystem. 

# Related GitHub Issues and PRs

- Ref: #146

# Checklist

- [x] I understand and agree to the [Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct/).
- [x] I have followed the [Style Guide for Python Code](https://peps.python.org/pep-0008/) as best as possible for the submitted code.
- [x] I have added **pytest** unit tests for any new functionality.

### CODE REVIEW

File: gt_extras/summary.py [hunk 1]
```diff
@@ -50,6 +57,33 @@ def gt_plt_summary(df: IntoDataFrame, title: str | None = None) -> GT:
...
+                "datetime": "#73a657",
+                "boolean": "#a65773",
+                "other": "black",
+            }
+
+            {
+                "string": "purple",
+                "numeric": "pink",
+            }
+
```
Comment (juleswg23) [inline gt_extras/summary.py:86]: Really appreciate the new docstring!

This is maybe a little nit-picky, but I think we can add an example with `new_color_mapping` to the main docstring examples section, and leave it out here. Instead you could describe what the dictionary is. Something like: `"A dictionary where the keys are ... and the values are ..."`

One way to generate the docs is by going into the root directory of the package and running `make docs-build` to build a local version of the documentation. Once you've done that once, you should be able to navigate into the docs directory (`cd docs`) and execute `quartodoc build; quarto preview`, to quickly see a live version of the API reference for this function.

File: gt_extras/summary.py [hunk 2]
```diff
@@ -50,6 +57,33 @@ def gt_plt_summary(df: IntoDataFrame, title: str | None = None) -> GT:
     title
         Optional title for the summary table. If `None`, defaults to "Summary Table".
 
+    hide_desc_stats
+        Boolean that allows the hiding of the Mean, Median, and SD columns. Defaults to 'False'.
+
+    add_mode
+        Boolean that allows the addition of a Mode column. Defaults to 'False'.
```
Comment (juleswg23) [inline gt_extras/summary.py:64]: We can take advantage of quarto to freshen up the docstrings. Using backticks (```` `False` ```` instead of ```` 'False' ```` or ```` `"string"` ```` instead of ```` "String" ````) will render the text as code.

**Comment by ScottFB101** [comment]
Of course! Happy to help, it's been good practice for me.

Alright I made all the changes, but will admit I'm struggling a bit figuring out the snapshot test (not something I've dealt with before), so I can't get 2 of the 500ish lines of `summary.py` to get coverage using pytest.

**Comment by juleswg23** [comment]
> I can't get 2 of the 500ish lines of `summary.py` to get coverage using pytest.

This is getting close! The snapshot test is great, it's not necessarily supposed to hit every branch, just to check a standard example's output. If you run `make test-update` and commit the `test_summary.ambr` file in the `tests/__snapshots/` directory, those snapshot tests should pass with all future as you have them now in all future `make test` executions.

To get coverage on the un-hit lines (for example, I am seeing the below line isn't covered by the snapshot) you can write tests similar to the rest of the ones in `test_summary.py`. So for example, to test the below branch:

```python
# Limiting the number of modes displayed to two at maximum
elif len(mode_val) > 2:
    mode_val = "Greater than 2 Modes"
# Converting to string, then listing together
else:
    mode_val = ", ".join(str(i) for i in sorted(mode_val.to_list()))
```

We can use tests like:

``` python

@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_two_modes(DataFrame):
    df = DataFrame({"numeric": [1, 1, 2, 2, 3]})

    result = gt_plt_summary(df, add_mode=True)
    html = result.as_raw_html()

    assert '<td class="gt_row gt_left">1, 2</td>' in html

@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_greater_than_two_modes(DataFrame):
    df = DataFrame({"numeric": [1, 1, 2, 2, 3, 4, 4]})

    result = gt_plt_summary(df, add_mode=True)
    html = result.as_raw_html()

    assert '<td class="gt_row gt_left">Greater than 2 Modes</td>' in html
```

I know it's not the flashiest to do, but it would be nice to have these kinds of specific tests for each individual new parameter, and then optionally any edge cases you might want to include. (For example, establishing behavior that is expected when `add_mode=True` and `show_desc_stats=False`.)

**Comment by ScottFB101** [comment]
> > I can't get 2 of the 500ish lines of `summary.py` to get coverage using pytest.
> 
> ```python
> # Limiting the number of modes displayed to two at maximum
> elif len(mode_val) > 2:
>     mode_val = "Greater than 2 Modes"
> # Converting to string, then listing together
> else:
>     mode_val = ", ".join(str(i) for i in sorted(mode_val.to_list()))
> ```
> 
> We can use tests like:
> 
> ```python
> @pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
> def test_gt_plt_summary_two_modes(DataFrame):
>     df = DataFrame({"numeric": [1, 1, 2, 2, 3]})
> 
>     result = gt_plt_summary(df, add_mode=True)
>     html = result.as_raw_html()
> 
>     assert '<td class="gt_row gt_left">1, 2</td>' in html
> 
> @pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
> def test_gt_plt_summary_greater_than_two_modes(DataFrame):
>     df = DataFrame({"numeric": [1, 1, 2, 2, 3, 4, 4]})
> 
>     result = gt_plt_summary(df, add_mode=True)
>     html = result.as_raw_html()
> 
>     assert '<td class="gt_row gt_left">Greater than 2 Modes</td>' in html
> ```

Unfortunately Pandas handles the Mode column as an object. I guess there's no native string column type? Polars can handle string columns, so the alignment in the data cell varies depending on dataframe type. For example, Pandas results in "gt_right", Polars in "gt_left".

File: gt_extras/summary.py [hunk 3]
```diff
@@ -113,13 +134,54 @@ def gt_plt_summary(df: IntoDataFrame, title: str | None = None) -> GT:
...
     Note
     ---------
     The datatype (dtype) of each column in your dataframe will determine the classified type in the
     summary table. Keep in mind that sometimes pandas or polars have differing behaviors with
     datatypes, especially when null values are present.
     """
-    summary_df = _create_summary_df(df)
+    if new_color_mapping:
+        global COLOR_MAPPING
+        COLOR_MAPPING.update(new_color_mapping)
```
[the remark this answers is not in the document]
Comment (juleswg23) [reply gt_extras/summary.py:180]: I don't think it solves the root of the problem.

If I run these three in succession, the third one is colored red, rather than orange.

```python
import pandas as pd
from gt_extras import gt_plt_summary

df = pd.DataFrame({"numeric": [1, 1, 2, 2, 3]})
gt_plt_summary(df)
```

``` python
df = pd.DataFrame({"numeric": [1, 1, 2, 2, 3]})
gt_plt_summary(df, new_color_mapping={"numeric":"red"})
```

``` python
df = pd.DataFrame({"numeric": [1, 1, 2, 2, 3]})
gt_plt_summary(df)
```

image

**Comment by juleswg23** [comment]
> Unfortunately, Pandas handles the Mode column as an object. I guess there's no native string column type? Polars can handle string columns, so the alignment in the data cell varies depending on dataframe type. For example, Pandas results in "gt_right", Polars in "gt_left".
> 
> For the time being, I strictly checked that the correct values were in the data cell, but didn't check alignment.

Yeah, I'm seeing this too. I'd guess there's some workaround within the dataframe, but we could supersede it and set the following:

``` python
gt = gt.cols_align(align="right", columns="Mode")
```

Just note that's not going to work if the mode column isn't present.

I also don't think it's that bad to leave it as is. In that case, the user can edit the table if they see fit.

File: gt_extras/summary.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ScottFB101) [reply gt_extras/summary.py:180]: Must of commented in the wrong spot originally, so there may be a similar response in this PR thread.

I could make a Pytest fixture that resets the color mapping to the preset from the `summary.py` script prior to each test.
